### DISCUSSION TITLE
Вопрос про расширение БД

### DISCUSSION THREAD

**Discussion by AnarkiQ3** [erthink/libmdbx#355] [Q&A] [answered]
Ситуация такая: у нас есть одно приложение писатель и несколько приложений читателей. 
У читателей хранится MDBX_env и MDBX_txn, в конце транзакций вызывается mdbx_txn_reset. Изначально под Windows мы с помощью mdbx_env_set_geometry сразу выделяли в писателе большую базу определенного размера и удаляли по таймеру старые записи, однако столкнулись с тем, что могли преждевременно достигнуть потолка и удалить что-либо не представлялось возможным по причине переполнения базы. 
Сейчас мы задали верхний предел выше минимального размера бд и столкнулись с тем, что после расширения при попытке открыть транзакцию возвращается MDBX_MAP_RESIZED и необходимо переоткрывать MDBX_env, но даже делая это существует вероятность открыть транзакцию чтения и уже после этого при вызове mdbx_cursor_open получить ошибку MDBX_CORRUPTED, вероятно после очередного расширения бд.
Вопрос собственно в следующем:
Есть ли какой-то определенный алгоритм, по которому необходимо работать читателям в такой ситуации? (при расширении базы данных)

**Comment by erthink** [comment, marked as answer]
1. Вы не верно используете API.
   - Изменение верхнего предела размера БД требует изменения размера mmap-отображения, т.е. изменения резервирования адресного пространства в каждом работающим с БД процессе. Это очень затратный процесс, который в Windows требует снятия и повторного создания отображения.
   - Если один из процессов меняет размер отображения, то все остальные должны сделать тоже самое для возможности работы с БД.
   - Если внутри какого-то процесса адресное пространство за образом БД в памяти занято, то он не сможет изменить для себя размер видимой части БД без её пере-открытия. Соответственно вы ловите ошибки.

2. Вам нужно _один_ раз задать требуемую геометрию БД с достаточно большим верхним пределом размера и подходящим шагом увеличения.
   - Шаг уменьшения может быть нулевым, но если хотите чтобы БД могла уменьшится, то его следует задать в два раза больше шага увеличения.
Однако, стоит отметить, что из-за системных ограничений Windows уменьшения размера БД возможно только последним/единственным работающим с БД процессом, т.е. уменьшение скорее всего произойдет только при закрытии БД.
   - Нижний предел размера и текущий размер вы можете установить в минимальное значение, либо заранее выделить место на диске.

3. Чистку БД от старых записей желательно делать не только по устареванию, но и при малом кол-ве свободного места. Т.е. наверное, лучше всего, удалять старые записи непосредственно в пишущей транзакции, перед добавлением новых данных.

4. Тем не менее, в описываемой вами ситуации, при невозможности расширить размер отображения, для читателя должна генерировать ошибка `MDBX_MAP_RESIZED`, а не `MDBX_CORRUPTED`. Поэтому мне бы хотелось получить более подробное описание сценария.

**Comment by AnarkiQ3** [comment]
Геометрия задается только один раз, при запуске писателя, хотя конечно возможно, что писатель может запускаться позже, чем читатели, но обычно изменения геометрии не требуется и она задается такой же

Насчет удаления по малому кол-ву свободного места, хотелось бы понять, какое количество старых записей необходимо удалять (т.е. нужно удалять столько же записей, сколько собираемся записать или больше, а так же стоит ли удалять их в каждой транзакции или стоит удалить в одной транзакции больше записей чтобы какое-то время не беспокоиться об этом), а так же как правильно определять текущее заполнение базы?

Насчет `MDBX_CORRUPTED`, это происходит после успешного выполнения `mdbx_txn_renew` (где  обычно и генерируется ошибка `MDBX_MAP_RESIZED`), и `mdbx_dbi_open`, а генерируется уже в `mdbx_cursor_open`, в котором нигде не генерируется `MDBX_MAP_RESIZED`. Пока воспроизвести это не удалось, но я связываю это с расширением бд потому что совпадает интервал времени между предыдущими ошибками `MDBX_MAP_RESIZED` в `mdbx_txn_renew` и этой ошибкой.

**Comment by erthink** [comment]
> Геометрия задается только один раз, при запуске писателя, хотя конечно возможно, что писатель может запускаться позже, чем читатели, но обычно изменения геометрии не требуется и она задается такой же.

Задавайте геометрию _один раз при создании БД_, до запуска основных процессов работающих с БД, а в самих процессах открывайте готовую БД (с возвратам ошибки при её отсутствии) . Остальные варианты не могут надежно предотвратить изменение размера отображения "на ходу" и следовательно гарантировать успеха во всех случаях.

> Насчет удаления по малому кол-ву свободного места, хотелось бы понять, какое количество старых записей необходимо удалять (т.е. нужно удалять столько же записей, сколько собираемся записать или больше, а так же стоит ли удалять их в каждой транзакции или стоит удалить в одной транзакции больше записей чтобы какое-то время не беспокоиться об этом), а так же как правильно определять текущее заполнение базы?

Ну есть же очевидные вещи - если удалять меньше чем добавлять, то рано или поздно БД заполнится, и чем меньше размер БД, тем проще системе с ней работать.

Пакетное удаление в целом более выгодно, в том числе по объему записи на диск, но чем реже будет прореживание, тем менее равномерной будет нагрузка (объем БД и запись на диск). Тем не менее,  вы сами должны определиться с подробностями стратегии удаления старых записей.

Определить размер БД можно посредством API, см `struct MDBX_envinfo`, `struct MDBX_stat` и `struct MDBX_txn_info`.

**Comment by erthink** [comment]
@AnarkiQ3, закрываем issue или есть еще какие-то вопросы?

**Comment by AnarkiQ3** [comment]
Да, если вдруг получится воспроизвести ошибку MDBX_CORRUPTED, я напишу более подробно отдельно
